Indexing/Index.hpp

```diff
@@ -19,102 +19,124 @@
+ * See e.g. TermLiteralClause below for examples.
+ */
 namespace Indexing
 {
 using namespace Kernel;
 using namespace Lib;
 using namespace Saturation;
 
-
-/**
- * Class of objects which contain results of single literal queries.
- */
-template<class Unifier>
-struct LQueryRes
+struct LiteralClause 
 {
-  LQueryRes() {}
-  LQueryRes(Literal* l, Clause* c, Unifier unifier)
-  : literal(l), clause(c), unifier(std::move(unifier)) {}
+  Literal* const& key() const
+  { return literal; }
 
+private:
+  auto asTuple() const
+  { return std::make_tuple(clause->number(), literal->getId()); }
+public:
 
-  Literal* literal;
-  Clause* clause;
-  Unifier unifier;
+  IMPL_COMPARISONS_FROM_TUPLE(LiteralClause)
+
+  Literal* literal = nullptr;
+  Clause* clause = nullptr;
+
+  friend std::ostream& operator<<(std::ostream& out, LiteralClause const& self)
+  { return out << "{ " << outputPtr(self.clause) << ", " << outputPtr(self.literal) << " }"; }
+};
+
+template<class Value>
+struct TermWithValue {
+  TypedTermList term;
+  Value value;
+
+  TermWithValue() {}
+
+  TermWithValue(TypedTermList term, Value v)
+    : term(term)
+    , value(std::move(v))
+  {}
+
+  TypedTermList const& key() const { return term; }
+
+  auto asTuple() const
+  { return std::tie(term, value); }
+
+  IMPL_COMPARISONS_FROM_TUPLE(TermWithValue)
+
+  friend std::ostream& operator<<(std::ostream& out, TermWithValue const& self)
+  { return out << self.asTuple(); }
+};
+
+class TermWithoutValue : public TermWithValue<std::tuple<>> 
+{
+public:
+  TermWithoutValue(TypedTermList t) 
+    : TermWithValue(t, std::make_tuple()) 
+  { }
 };
 
-template<class Unifier>
-LQueryRes<Unifier> lQueryRes(Literal* l, Clause* c, Unifier unifier)
-{ return LQueryRes<Unifier>(l,c,std::move(unifier)); }
+struct TermLiteralClause 
+{
+  TypedTermList term;
+  Literal* literal = nullptr;
+  Clause* clause = nullptr;
+
+  TypedTermList const& key() const { return term; }
+
+  auto  asTuple() const
+  { return std::make_tuple(clause->number(), literal->getId(), term); }
+
+  IMPL_COMPARISONS_FROM_TUPLE(TermLiteralClause)
+
+  friend std::ostream& operator<<(std::ostream& out, TermLiteralClause const& self)
+  { return out << "("
+               << self.term << ", "
+               << self.literal
+               << outputPtr(self.clause)
+               << ")"; }
+};
 
 /**
  * Class of objects which contain results of term queries.
  */
-template<class Unifier>
-struct TQueryRes
+template<class Unifier, class Data>
+struct QueryRes
 {
-  TQueryRes() {}
-  TQueryRes(TermList t, Literal* l, Clause* c, Unifier unifier)
-  : term(t), literal(l), clause(c), unifier(std::move(unifier)) {}
-
-  TermList term;
-  Literal* literal;
-  Clause* clause;
   Unifier unifier;
+  Data const* data;
+
+  QueryRes() {}
+  QueryRes(Unifier unifier, Data const* data) 
+    : unifier(std::move(unifier))
+    , data(std::move(data)) {}
```

Comment:
Does it make sense to `std::move` a pointer? (Again, not 100% sure, but it looks strange to me.)

Reply:
No you're right, it doesn't make a difference. I just do always write a move in constructors just as sort of best-practice without thinking about the types.